src/JuliusSweetland.OptiKey.Core/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs

```diff
@@ -2376,45 +2372,131 @@ private void SelectVoice(string voice)
-                        HandleFunctionKeySelectionResult(singleKeyValue);
+                        Log.InfoFormat("StepListCommand triggering function key {0} press", actionFunctionKey);
+                        commandKeyValue = new KeyValue(actionFunctionKey);
+                        KeySelectionResult(commandKeyValue, multiKeySelection);
                     }
-                    else if (vStep.StartsWith("Keyboard>") || vStep.StartsWith("KeyboardAndReturn>"))
+                    else if (commandList.First().StartsWith("Keyboard>") || commandList.First().StartsWith("KeyboardAndReturn>"))
                     {
                         Enums.Keyboards keyboardEnum;
-                        bool vReturn = (vStep.StartsWith("KeyboardAndReturn>"));
-                        int vIndex = (vStep.StartsWith("KeyboardAndReturn>")) ? 18 : 9;
-                        if (System.Enum.TryParse(vStep.Substring(vIndex), out keyboardEnum))
+                        bool vReturn = (commandList.First().StartsWith("KeyboardAndReturn>"));
+                        int vIndex = (commandList.First().StartsWith("KeyboardAndReturn>")) ? 18 : 9;
+                        if (System.Enum.TryParse(commandList.First().Substring(vIndex), out keyboardEnum))
                         {
-                            ChangeKeyboardKeyValue keyValue = new ChangeKeyboardKeyValue(keyboardEnum, vReturn);
-                            ProcessChangeKeyboardKeyValue(keyValue);
+                            Log.InfoFormat("StepListCommand changing keyboard to {0}. Replace previous keyboard={1}", keyboardEnum, vReturn);
+                            commandKeyValue = new ChangeKeyboardKeyValue(keyboardEnum, vReturn);
                         }
                         else
                         {
-                            ChangeKeyboardKeyValue keyValue = new ChangeKeyboardKeyValue(vStep.Substring(vIndex), vReturn);
-                            ProcessChangeKeyboardKeyValue(keyValue);
+                            commandKeyValue = new ChangeKeyboardKeyValue(commandList.First().Substring(vIndex), vReturn);
+                            Log.InfoFormat("StepListCommand changing keyboard to string {0}. Replace previous keyboard={1}", commandKeyValue.String, vReturn);
                         }
+                        KeySelectionResult(commandKeyValue, multiKeySelection);
                     }
-                    else if (vStep.StartsWith("Text>"))
+                    else if (commandList.First().StartsWith("KeyToggle>"))
                     {
-                        keyboardOutputService.ProcessSingleKeyText(vStep.Substring(5));
+                        commandKeyValue = new KeyValue(commandList.First().Substring(10));
+                        Log.InfoFormat("StepListCommand: key toggle on {0} key", commandKeyValue.String);
+
+                        if (keyStateService.KeyDownStates[singleKeyValue].Value != KeyDownStates.Up)

```

Comment:
Could this be simplified to use the ProcessSingleKeyPress call with KeyPressType.Toggle?

Reply:
Perhaps. Though, sometimes I just want a universal all keys up when dealing with troublesome applications or remote computers.

Reply:
@AdamRoden I'm not saying we should remove any of KeyUp/KeyDown/Toggle functionality, but instead I was querying if the check on the KeyDownStates[singleKeyValue] needed to be done as there is a later check on the key state in the PublishService. However, I believe this is a different check as it is an OS level check to see what needs to be sent in the event to achieve the correct output, rather than a check on the key state within OptiKey. I think you can ignore my comment.